### DISCUSSION TITLE
zprint-filter-1.2.3

### DISCUSSION THREAD

**Discussion by kkinnear** [kkinnear/zprint#232] [General]
A fair number of fixes and changes in this release.  See the CHANGELOG.md for details, but notable new things are:

  * New `:range` configuration options: `:use-previous-!zprint?` and `:continue-after-!zprint-error?`.  These allow someone doing `:range` formatting on a file (string) which has `;!zprint` formatting directives to have the lines in the range formatted the same when doing just the range or the whole file.  `:use-previous-!zprint?` if true will scan all previous `;!zprint` directives prior to formatting the range. `:continue-after-!zprint-error?` if true will ignore errors in any `;!zprint` directives, and output any errors in the map returned by `:output {:range? true}`.

  * An alternative way to format metadata was created.  It is `:style :meta-alt`.  It is presently experiemental.  Issue #224.
 
 * `letfn` didn't indent its fns like fns.  Now it does.  And if they are multi-arity, the various arities don't hang, but rather flow. Issue #221.

  * The styles `:community` and `:rod` didn't compose, because `:guided` was not considered a "body" function type.  Now they do. Issue #223.

  * `zprint-file-str` was not thread-safe, for a variety of reasons. Now it is, and configuration is thread-safe as well.  Issue #226.

  * Figured out some Figwheel issues. If you use zprint with Figwheel, check out the CHANGELOG.md, or see issue #231.

<hr /><em>This discussion was created from the release <a href='https://github.com/kkinnear/zprint/releases/tag/1.2.3'>zprint-filter-1.2.3</a>.</em>

**Comment by arichiardi** [comment]
Hey @kkinnear I added `:meta-alt` 

```clojure
:style [:community :meta-alt :how-to-ns :binding-nl]
```

and it definitely addresses #224 - I think the only issue I see is this change:

```diff

-(deftest valid-ref?
-  (is (false? (sut/valid-ref? nil)))
+(deftest valid-ref? (is (false? (sut/valid-ref? nil)))
   (is (false? (sut/valid-ref? "")))
   (is (false? (sut/valid-ref? "i/do/not/start/with/a/slash")))
   (is (true? (sut/valid-ref? "/this/is/a/valid/ref")))
```

The  first`(is (false? ...` looks weird there

I then added this instead only for `deftest`

```clojure
{:style [:community :how-to-ns :binding-nl]
 :fn-map
 {"deftest" [:arg1-body {:style :meta-alt}]
 ...
}
```

That looked better, with only one little change that seems to happen all over the code base, a couple of examples:

```diff
 (t/deftest broadcast-to!
   (t/testing "cannot be performed in a transaction"
     (t/is (thrown? IllegalStateException
-                   (dosync (sut/broadcast-to! (mock-socket) (mock-topic) (mock-event)))))))
+                   (dosync (sut/broadcast-to! (mock-socket)
+                                              (mock-topic)
+                                              (mock-event)))))))
```

```diff
-          topic-a (sut/with-event-handlers (sut/->TopicController "topic/a" ::any [] [])
-                                           (sut/->EventHandler "topic.a/x" ::any))
+          topic-a (sut/with-event-handlers
+                   (sut/->TopicController "topic/a" ::any [] [])
+                   (sut/->EventHandler "topic.a/x" ::any))
```

```diff
 (deftest add-template-text-to-note
   (testing "note without value"
     (is (= {::model.note/id note-id ::model.note/value "foo"}
-           (sut/add-note-text-if-not-included {::model.note/id note-id} "foo"))
+           (sut/add-note-text-if-not-included {::model.note/id note-id}
+                                              "foo"))
         "it should add the new text"))
```

I found this worth mentioning cause there might be a bug lurking, let me know if you want me to try any alpha/beta version or have more questions!

**Reply by kkinnear** [reply]
Thanks for bringing this up.  I totally agree that `deftest` should be `:arg1-body` in `:style :meta-alt`, I'll make that change.  I think that the best intermediate approach would be: `{:fn-map {"deftest" [:arg1-body {:style :meta-base}]}}`.  Re-invoking `:meta-alt` (as you did, above) is probably not the best approach.

That said, I am unable to reproduce the other issue you mention.  Well, actually, it reproduces easily for me, since my default width is 80:
```
zprint.core=> (czprint i232c {:parse-string? true :style [:community :meta-alt :how-to-ns :binding-nl] :fn-map {"deftest" [:arg1-body {:style :meta-base}]}})
(deftest add-template-text-to-note
  (testing "note without value"
           (is (= {::model.note/id note-id, ::model.note/value "foo"}
                  (sut/add-note-text-if-not-included {::model.note/id note-id}
                                                     "foo"))
               "it should add the new text")))
nil

; If I make the width 90, then you get what you want:

zprint.core=> (czprint i232c {:parse-string? true :style [:community :meta-alt :how-to-ns :binding-nl] :fn-map {"deftest" [:arg1-body {:style :meta-base}]} :width 90})
(deftest add-template-text-to-note
  (testing "note without value"
           (is (= {::model.note/id note-id, ::model.note/value "foo"}
                  (sut/add-note-text-if-not-included {::model.note/id note-id} "foo"))
               "it should add the new text")))
```
So, I'm wondering if the changes in your formatting is somehow related to whatever your current width might be?  I've looked, hard, and there isn't anything in the `:style`s that affects the width.  

If you give me your complete configuration (including whatever your current width is) I'll try to reproduce this as well.  Thanks!

**Reply by arichiardi** [reply]
Thank @kkinnear for your answer. I've moved the config to 

`  "deftest" [:arg1-body {:style :meta-base}]`

and that did the trick.

The weird thing is that I cannot reproduce the width problem anymore and I am baffled as to why that happened, see #224 on actual test cases
